File: 2024/day19/day19.nim
import std/[strutils, tables, sequtils]

var
    pats = newSeq[string]()
    towels = newSeq[string]()

var l = ""
var n2 = false
for l in lines("input.txt"):
    if l == "":
        n2 = true
        continue
    if n2:
        towels.add(l)
    else:
        let l2 = l.split(", ")
        for i in l2:
            pats.add(i)

var memo = initTable[string, int]()

proc part1(s: string): int = 
    if len(s) == 0: 
        return 1
    var sum  = 0
    for p in pats:
        if s.startsWith(p):
            let ns = s[len(p)..<len(s)]
            if memo.hasKey(ns):
                sum += memo[ns]
            else:
                memo[ns] = part1(ns)
                sum += memo[ns]
    return sum

var count = 0
for t in towels:
    clear(memo)
    if part1(t) > 0:
        inc(count,1)
echo count

File: 2024/day19/day19.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

var (
	pats   map[string]any = make(map[string]any)
	towels []string
	memo   map[string]int64 = make(map[string]int64)
)

func load(file string) {
	f, _ := os.Open(file)
	defer f.Close()
	sc := bufio.NewScanner(f)
	isfirst := true
	for sc.Scan() {
		t := sc.Text()
		if t == "" {
			isfirst = false
			continue
		}
		if isfirst {
			for _, v := range strings.Split(t, ", ") {
				pats[v] = nil
			}
		} else {
			towels = append(towels, t)
		}
	}
}

func part1(s string) int64 {
	if len(s) == 0 {
		return 1
	}
	var sum int64 = 0
	for p := range pats {
		if strings.HasPrefix(s, p) {
			ns := s[len(p):]
			if v, ok := memo[ns]; ok {
				sum += v
			} else {
				memo[ns] = part1(ns)
				sum += memo[ns]
			}
		}
	}
	return sum
}

func solvePart1() {
	load("input.txt")
	count := 0
	for _, t := range towels {
		clear(memo)
		if part1(t) > 0 {
			count++
		}
	}
	println(count)
}

func part2(towel string) int64 {
	if len(towel) == 0 {
		return 1
	}
	var sum int64 = 0
	for p := range pats {
		if strings.HasPrefix(towel, p) {
			ns := towel[len(p):]
			if v, ok := memo[ns]; ok {
				sum += int64(v)
			} else {
				memo[ns] = part2(ns)
				sum += memo[ns]
			}
		}
	}
	return sum
}

// sovle part2
func main() {
	load("input.txt")
	count := int64(0)
	for _, t := range towels {
		clear(memo)
		count += part2(t)
	}
	fmt.Println(count)
}
